### PR TITLE
Add overwritable HasTypes to support generic dtype conversions

### PR DESCRIPTION
**Pull request by junjihashimoto**
This PR provides HasTypes typeclass for general dtype conversions.
This HasTypes typeclass is similar to HasTypes in generic-lens, but the behavior of this PR's HasTypes can be overridden simply by defining an instance. 
This PR hasn't changed Parameterized typeclass at all.  (It is  not trying to force a merge of Parameterized and GTravelable typeclass like https://github.com/hasktorch/hasktorch/pull/505.)

### CODE REVIEW

**Review by tscholak — APPROVED** [review]
thanks @junjihashimoto, this is a step in the right direction!

**Comment by junjihashimoto** [comment]
Thank you for your reviewing!

**Comment by austinvhuang** [comment]
@junjihashimoto - can we either have the changed mnist-mlp be a seperate example (eg mnist-mixed-precision)? mnist-mlp will be many peoples' "hello world" onboarding experience so i think the fewer concepts they need to be familiar with, the better.

Besides that I think it looks good!

**Comment by junjihashimoto** [comment]
@austinvhuang Thank you for your reviewing! 
The separation of the codes is done.

**Review by austinvhuang — APPROVED** [review]
LGTM!

File: examples/mnist-mlp/Main.hs [hunk 1]
```diff
@@ -74,24 +78,54 @@ displayImages model (testImg, testLabel) = do
...
+              then tensor'
+              else toType Float tensor'
+       in tensor''
+
 main :: IO ()
 main = do
   deviceStr <- try (getEnv "DEVICE") :: IO (Either SomeException String)
-  let localDevice = case deviceStr of
+  dtypeStr <- try (getEnv "DTYPE") :: IO (Either SomeException String)
+  let device' = case deviceStr of
```
[the remark this answers is not in the document]
Comment (junjihashimoto) [reply examples/mnist-mlp/Main.hs]: I'm thinking these type-classes for ptlighting-like interfaces.
https://gist.github.com/junjihashimoto/6d393fbc61920afee3a8f0e48b89e9a2

[the remark this answers is not in the document]
Comment (junjihashimoto) [reply examples/mnist-mlp/Main.hs]: I am also thinking about monadic interfaces to lift devices and dtypes. There is no code yet.

**Comment by junjihashimoto** [comment]
After merge this pr, I want to see the end user response.

**Pull request merged**